### PR TITLE
Nvidia GPU utilization

### PR DESCRIPTION
**Pull request by Cyclenerd**
I cleaned up `nvidia_gpu_` a bit and implemented ShellCheck recommendations.

Then I implemented a new graph for GPU utilization. Tested with two GTX 1080 using driver version 390.25.

### CODE REVIEW

File: plugins/gpu/nvidia_gpu_ [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/env bash
```
Comment (sumpfralle) [inline plugins/gpu/nvidia_gpu_]: Please stick with `/bin/bash` for now.
The interpreter detection of our tests currently fail on the above `env`. Thus for now `/bin/bash` is the way.
Please adjust locally, if necessary.

File: plugins/gpu/nvidia_gpu_ [hunk 2]
```diff
@@ -87,95 +86,110 @@ if [ "$1" = "suggest" ]; then
...
 	echo "power"
+	echo "utilization"
 	exit 0
 fi
 
 # Get number of GPUs
-nGpusOutput=`$nvSmiExec -L`
-nGpus=`echo "$nGpusOutput" | wc -l`
-if [ $nGpus -eq 0 ]; then
+nGpusOutput=$($nvSmiExec -L)
```
Comment (sumpfralle) [inline plugins/gpu/nvidia_gpu_]: You should quote `$nvSmiExec` here, too, I guess?
Or could it be a multi-token thing?

File: plugins/gpu/nvidia_gpu_ [hunk 3]
```diff
@@ -87,95 +86,110 @@ if [ "$1" = "suggest" ]; then
...
+nGpus=$(echo "$nGpusOutput" | wc -l)
+if [ "$nGpus" -eq 0 ]; then
 	# Exit if no GPUs found
 	echo "No NVIDIA GPUs detected. Exiting."
 	exit 1
 fi
 
 # Get full output from nvidia-smi
-smiOutput=`$nvSmiExec -q`
+smiOutput=$($nvSmiExec -q)
```
Comment (sumpfralle) [inline plugins/gpu/nvidia_gpu_]: Quoting `$nvSmiExec`?

File: plugins/gpu/nvidia_gpu_ [hunk 4]
```diff
@@ -87,95 +86,110 @@ if [ "$1" = "suggest" ]; then
...
-			while [ $nGpusCounter -lt $nGpus ]
+			while [ "$nGpusCounter" -lt "$nGpus" ]
 			do
-				gpuName=`echo "$nGpusOutput" | sed -n $(( $nGpusCounter + 1 ))p | cut -d \( -f 1`
-				echo "temp${nGpusCounter}.warning ${warning:-75}"
-				echo "temp${nGpusCounter}.critical ${critical:-95}"
-				echo "temp${nGpusCounter}.info Temperature information for $gpuName"
-				: $(( nGpusCounter = $nGpusCounter + 1 ))
-			done 
+				gpuName=$(echo "$nGpusOutput" | sed -n $((nGpusCounter+1))p | cut -d \( -f 1)
```
Comment (sumpfralle) [inline plugins/gpu/nvidia_gpu_]: Please quote `$((nGpusCounter+1))p`.

File: plugins/gpu/nvidia_gpu_ [hunk 5]
```diff
@@ -87,95 +86,110 @@ if [ "$1" = "suggest" ]; then
...
+			gpusTotalMemOutput=$(echo "$smiOutput" | grep -v BAR1 | grep -A 3 "Memory Usage" | grep "Total" | cut -d : -f 2 | tr -d ' ')
 			gpusTotalMem=''
 			nGpusCounter=0
-			while [ $nGpusCounter -lt $nGpus ]
+			while [ "$nGpusCounter" -lt "$nGpus" ]
 			do
-				gpuName=`echo "$nGpusOutput" | sed -n $(( $nGpusCounter + 1 ))p | cut -d \( -f 1`
-				echo "mem${nGpusCounter}.info Memory information for $gpuName"
-				gpuMem=`echo "$gpusTotalMemOutput"| sed -n $(( $nGpusCounter + 1 ))p`
+				gpuName=$(echo "$nGpusOutput" | sed -n $((nGpusCounter+1))p | cut -d \( -f 1)
```
Comment (sumpfralle) [inline plugins/gpu/nvidia_gpu_]: Please quote

File: plugins/gpu/nvidia_gpu_ [hunk 6]
```diff
@@ -87,95 +86,110 @@ if [ "$1" = "suggest" ]; then
...
 			echo "graph_info Fan speed of NVIDIA GPUs using driver version $driverVersion"
 			nGpusCounter=0
-			while [ $nGpusCounter -lt $nGpus ]
+			while [ "$nGpusCounter" -lt "$nGpus" ]
 			do
-				gpuName=`echo "$nGpusOutput" | sed -n $(( $nGpusCounter + 1 ))p | cut -d \( -f 1`
-				echo "fan${nGpusCounter}.info Fan information for $gpuName"
-				: $(( nGpusCounter = $nGpusCounter + 1 ))
-				done 
+				gpuName=$(echo "$nGpusOutput" | sed -n $((nGpusCounter+1))p | cut -d \( -f 1)
```
Comment (sumpfralle) [inline plugins/gpu/nvidia_gpu_]: Please quote.

Maybe you want to move these identical evaluations into a function? (e.g. `get_cpu_name_by_id`)

File: plugins/gpu/nvidia_gpu_ [hunk 7]
```diff
@@ -87,95 +86,110 @@ if [ "$1" = "suggest" ]; then
...
 			echo 'graph_category sensors'
 			echo "graph_info power consumption of NVIDIA GPUs using driver version $driverVersion"
 			nGpusCounter=0
-			while [ $nGpusCounter -lt $nGpus ]
+			while [ "$nGpusCounter" -lt "$nGpus" ]
 			do
-				gpuName=`echo "$nGpusOutput" | sed -n $(( $nGpusCounter + 1 ))p | cut -d \( -f 1`
-				echo "power${nGpusCounter}.info power consumption of $gpuName"
-				: $(( nGpusCounter = $nGpusCounter + 1 ))
+				gpuName=$(echo "$nGpusOutput" | sed -n $((nGpusCounter+1))p | cut -d \( -f 1)
```
Comment (sumpfralle) [inline plugins/gpu/nvidia_gpu_]: Please quote.

File: plugins/gpu/nvidia_gpu_ [hunk 8]
```diff
@@ -87,95 +86,110 @@ if [ "$1" = "suggest" ]; then
...
+		utilization)
+			echo 'graph_title GPU utilization'
+			echo 'graph_args -l 0 -u 100'
+			echo 'graph_vlabel %'
+			echo 'graph_category system'
+			echo "graph_info GPU utilization of NVIDIA GPUs using driver version $driverVersion"
+			nGpusCounter=0
+			while [ "$nGpusCounter" -lt "$nGpus" ]
+			do
+				gpuName=$(echo "$nGpusOutput" | sed -n $((nGpusCounter+1))p | cut -d \( -f 1)
```
Comment (sumpfralle) [inline plugins/gpu/nvidia_gpu_]: Please quote.

File: plugins/gpu/nvidia_gpu_ [hunk 9]
```diff
@@ -185,11 +199,11 @@ if [ "$1" = "config" ]; then
 
 	# Common stuff for all quantities
 	nGpusCounter=0
-	while [ $nGpusCounter -lt $nGpus ]
+	while [ "$nGpusCounter" -lt "$nGpus" ]
 	do
-		gpuName=`echo "$nGpusOutput" | sed -n $(( $nGpusCounter + 1 ))p | cut -d \( -f 1`
+		gpuName=$(echo "$nGpusOutput" | sed -n $((nGpusCounter+1))p | cut -d \( -f 1)
```
Comment (sumpfralle) [inline plugins/gpu/nvidia_gpu_]: Please quote.

File: plugins/gpu/nvidia_gpu_ [hunk 10]
```diff
@@ -200,27 +214,30 @@ fi
...
+		usedMemGpus=$(echo "$smiOutput" | grep -v BAR1 | grep -A 3 "Memory Usage" | grep "Used" | cut -d : -f 2 | cut -d ' ' -f 2)
 		valueGpus=''
 		nGpusCounter=0
-		while [ $nGpusCounter -lt $nGpus ]
+		while [ "$nGpusCounter" -lt "$nGpus" ]
 		do
-			totalMemGpu=`echo "$totalMemGpus" | sed -n $(( $nGpusCounter + 1 ))p`
-			usedMemGpu=`echo "$usedMemGpus" | sed -n $(( $nGpusCounter + 1 ))p`
-			percentMemUsed=$(( $usedMemGpu * 100 / $totalMemGpu ))
+			totalMemGpu=$(echo "$totalMemGpus" | sed -n $((nGpusCounter+1))p)
```
Comment (sumpfralle) [inline plugins/gpu/nvidia_gpu_]: Please quote (also the following lines).

File: plugins/gpu/nvidia_gpu_ [hunk 11]
```diff
@@ -232,12 +249,9 @@ case $name in
 
 # Print requested value
 nGpusCounter=0
-while [ $nGpusCounter -lt $nGpus ]
+while [ "$nGpusCounter" -lt "$nGpus" ]
 do
-	value=`echo "$valueGpus" | sed -n $(( $nGpusCounter + 1 ))p`
+	value=$(echo "$valueGpus" | sed -n $((nGpusCounter+1))p)
```
Comment (sumpfralle) [inline plugins/gpu/nvidia_gpu_]: Please quote.

**Comment by Cyclenerd** [comment]
Hi,

I have successfully revised it.  

Quotation`$nvSmiExec` makes sense. Can be handed over by the user.

Quoting in calculations and is not necessary and does not work (`syntax error: operand expected`):
```
...
2018/02/23-19:10:04 [12604] 	/etc/munin/plugins/nvidia_gpu_mem: line 163: "nGpusCounter"+1: syntax error: operand expected (error token is ""nGpusCounter"+1")
2018/02/23-19:10:04 [12604] 	/etc/munin/plugins/nvidia_gpu_mem: line 165: "nGpusCounter"="nGpusCounter"+1: syntax error: operand expected (error token is ""nGpusCounter"="nGpusCounter"+1")
...
```

Quoting `$nGpusCounter` is also not necessary. Variable is guaranteed a number.

ShellCheck provides no more warnings. 

Changes again successfully tested with the same hardware and drivers.

Best regards
Nils

**Comment by sumpfralle** [comment]
> Quoting in calculations and is not necessary and does not work (syntax error: operand expected)

I am sorry: I did not express myself clearly. I meant the following: `sed -n "$((nGpusCounter+1))p"`.
Your are right: it is not strictly necessary, since it is an arithmetical expression and it could never contain whitespace. Shellcheck noticed that and thus did not complain.
Personally I would always quote these situations, too. I simply want to save myself from having to think about potential quoting issues again, when I need to look at the code next time :)

The same goes for the quotes that you removed (`[ $nGpusCounter -lt "$nGpus" ]`): shellcheck is clever and notices, that this variable is always the result of arithmetical expressions - thus it does not complain about the missing quotes.
Here I would also always quote the expression - just for saving the next reader the time for checking it again.
Anyway: it is shellcheck-clean - thus everything is fine as it is.

Btw.: the only line of code in this script that requires "bash" (instead of `/bin/sh`) is this one:
```shell
valueGpus="${valueGpus}${percentMemUsed}"$'\n'
```
Moving the `\n` before the closing `"` would remove the need for the bash expression `$'\n'`.

Anyway: if you are fine with the current state, I will just merge it. Should I?

**Comment by Cyclenerd** [comment]
Thank you for the explanation. I would like to leave the program code that way. Please merge. Many Thanks.

**Comment by sumpfralle** [comment]
Great - thank you!

**Pull request merged**